sheerwater_benchmarking/data/stage_iv.py

```diff
@@ -0,0 +1,123 @@
+"""Stage IV precipitation data product."""
+import numpy as np
+import xarray as xr
+from dateutil import parser
+
+from sheerwater_benchmarking.utils import apply_mask, cacheable, clip_region, dask_remote, regrid, roll_and_agg
+
+
+@dask_remote
+@cacheable(data_type='array',
+           cache_args=['year', 'grid'],
+           chunking={'lat': 300, 'lon': 300, 'time': 365})
+def stage_iv_gridded(year, grid):
+    """Regridded Stage IV precipitation data by year.
+
+    Reads data from USGS OSN zarr store.
+    Data is available from 2002-present.
+    """
+    # Open the zarr store
+    store = 'https://usgs.osn.mghpcc.org/mdmf/gdp/stageiv_combined.zarr/'
+    ds = xr.open_dataset(store, engine='zarr', chunks={})
+
+    # Select the year
+    ds = ds.sel(time=slice(f"{year}-01-01", f"{year}-12-31"))
+
+    # Ensure consistent naming
+    ds = ds.rename(
+        {'Total_precipitation_surface_1_Hour_Accumulation': 'precip'},
+    )
+
+    # chirps is in "units": "mm/day"
+    # stage iv is in "units": "kg m-2" and is measured hourly
+    # 1 kg/m^2 = 1 mm, so just need to sum to daily values
+    ds = ds.resample(time='1D').sum()
+    ds.attrs.update(units='mm/day')
+
+    # Fix coordinates and dimensions
+    # Data is in polar stereographic projection with 2D lat/lon coordinates
+    # Need to interpolate to a regular lat/lon grid before regridding
+
+    # Get min/max bounds of the data
+    lat_min = ds.latitude.min().values
+    lat_max = ds.latitude.max().values
+    lon_min = ds.longitude.min().values
+    lon_max = ds.longitude.max().values
+
+    # Create regular lat/lon coordinates at approximately the same resolution
+    n_lat = ds.dims['y']
+    n_lon = ds.dims['x']
+    new_lats = np.linspace(lat_min, lat_max, n_lat)
+    new_lons = np.linspace(lon_min, lon_max, n_lon)
+
+    # Interpolate to the new regular grid
+    ds = ds.interp(
+        y=new_lats,
+        x=new_lons,
+        method='linear'
+    )
+
+    # Drop any existing coordinates
+    ds = ds.reset_coords(drop=True)
+
+    # Rename dimensions and assign final coordinates
+    ds = ds.rename({'y': 'lat', 'x': 'lon'})
+    ds = ds.assign_coords({
+        'lat': ('lat', new_lats),
+        'lon': ('lon', new_lons)
+    })
+
+    ds = ds.drop_vars(['crs'])
+
+    # regrid
+    ds = regrid(ds, grid, base='base180', method='conservative')
+
+    return ds
+
+@dask_remote
+@cacheable(data_type='array',
+           timeseries=['time'],
+           cache_args=['grid', 'agg_days'],
+           chunking={'lat': 300, 'lon': 300, 'time': 365})
+def stage_iv_rolled(start_time, end_time, agg_days, grid):
+    """Stage IV rolled and aggregated."""
+    years = range(parser.parse(start_time).year, parser.parse(end_time).year + 1)
+
+    datasets = []
+    for year in years:
+        # NOTE: filepath_only is used by the cacheable decorator
```

Comment:
Yes, this note is correct - you can remove the comment, or refine to something like: lazily gather the filepath objects for a multi-year load

Reply:
I removed the comment. I had just added it because it was a little unclear where this was coming from in the beginning.